src/file_walker.rs

```diff
@@ -0,0 +1,670 @@
+    pub fn max_depth(mut self, depth: Option<usize>) -> Self {
+        self.max_depth = depth;
+        self
+    }
+
+    /// Add additional paths to traverse
+    pub fn add_path<P: AsRef<Path>>(mut self, path: P) -> Self {
+        self.paths.push(path.as_ref().to_path_buf());
+        self
+    }
+
+    /// Collect all paths matching a filter predicate
+    pub fn collect<F>(self, filter: F) -> Vec<PathBuf>
+    where
+        F: Fn(&DirEntry) -> bool + Sync + Send,
+    {
+        let follow_links = self.follow_links;
+        let max_depth = self.max_depth;
+
+        // Create a thread pool with a specific number of threads and names
+        let pool = rayon::ThreadPoolBuilder::new()
+            .num_threads(num_cpus::get() - 1)
+            .build()
+            .unwrap();
+
+        // Execute a task within the thread pool and collect results
+        pool.install(|| {
+            // Process all root paths in parallel and collect results
+            if self.paths.len() == 1 {
+                // fast path
+                Self::walk_dir_parallel_collect(&self.paths[0], 0, &filter, follow_links, max_depth)
+            } else {
+                self.paths
+                    .into_par_iter()
+                    .flat_map(|path| {
+                        Self::walk_dir_parallel_collect(&path, 0, &filter, follow_links, max_depth)
+                    })
+                    .collect()
+            }
+        })
+    }
+
+    /// Static version of walk_dir_parallel_collect to collect paths matching a filter
+    fn walk_dir_parallel_collect<F>(
+        path: &Path,
+        depth: usize,
+        filter: &F,
+        follow_links: bool,
+        max_depth: Option<usize>,
+    ) -> Vec<PathBuf>
+    where
+        F: Fn(&DirEntry) -> bool + Sync + Send,
+    {
+        // Check depth limit
+        if let Some(max_depth) = max_depth {
+            if depth > max_depth {
+                return Vec::new();
+            }
+        }
+
+        let entries = match fs::read_dir(path) {
+            Ok(entries) => entries,
+            Err(e) => {
+                debug!("Error reading directory {:?}: {}", path, e);
+                // If we can't read the directory, return an empty vector
+                return Vec::new();
+            }
+        };
+
+        // Use par_bridge() to parallelize the directory iterator efficiently
+        entries
+            .par_bridge()
+            .filter_map(|entry_result| {
+                match entry_result {
+                    Ok(entry) => {
+                        // Get file type first before moving entry
+                        let file_type = entry.file_type();
+                        let entry_path = entry.path();
+                        // Check if this entry matches the filter
+                        let matched = filter(&entry);
+
+                        // Check if it's a directory for recursion
+                        let is_dir = file_type.is_ok_and(|file_type| {
```

Comment:
We should track the already visited directories in order to avoid to revisit them in case of a symlink somewhere inside a dir pointing on this dir.
It shouldn't happen in real life... but who knows, it's probably better to be preventive here.